chains/evm/contracts/pools/USDC/USDCTokenPoolProxy.sol

```diff
@@ -0,0 +1,308 @@
+  /// @param remoteChainSelectors The remote chain selectors to update the lock or burn mechanism for.
+  /// @param mechanisms The new lock or burn mechanisms for the given remote chain selectors.
+  /// @dev Only callable by the owner.
+  function updateLockOrBurnMechanisms(
+    uint64[] calldata remoteChainSelectors,
+    LockOrBurnMechanism[] calldata mechanisms
+  ) external onlyOwner {
+    if (remoteChainSelectors.length != mechanisms.length) {
+      revert MismatchedArrayLengths();
+    }
+
+    for (uint256 i = 0; i < remoteChainSelectors.length; ++i) {
+      if (!isSupportedChain(remoteChainSelectors[i])) {
+        revert TokenPool.NonExistentChain(remoteChainSelectors[i]);
+      }
+
+      s_lockOrBurnMechanism[remoteChainSelectors[i]] = mechanisms[i];
+      emit LockOrBurnMechanismUpdated(remoteChainSelectors[i], mechanisms[i]);
+    }
+  }
+
+  /// @notice Check if the releaseOrMintIn struct is an inflight message from a legacy pool that did not utilize a
+  /// message transmitter proxy.
+  /// @param releaseOrMintIn The releaseOrMintIn struct to check for a legacy inflight message.
+  /// @return True if the releaseOrMintIn struct is an inflight message from a legacy pool that did not utilize a
+  /// message transmitter proxy, false otherwise.
+  function _checkForLegacyInflightMessages(
+    Pool.ReleaseOrMintInV1 calldata releaseOrMintIn
+  ) internal view virtual returns (bool) {
+    // Cache the legacy pool address to avoid multiple SLOADs.
+    address legacyPool = s_pools.legacyCctpV1Pool;
+
+    // If the legacy pool without a proxy is not set, then there is no need to check the destinationCaller.
+    if (legacyPool == address(0)) {
+      return false;
+    }
+
+    USDCTokenPool.MessageAndAttestation memory msgAndAttestation =
+      abi.decode(releaseOrMintIn.offchainTokenData, (USDCTokenPool.MessageAndAttestation));
+
+    bytes32 destinationCallerBytes32;
+    bytes memory messageBytes = msgAndAttestation.message;
+    assembly {
+      // destinationCaller is a 32-byte word starting at position 84 in messageBytes body, so add 32 to skip the 1st word
+      // representing bytes length.
+      destinationCallerBytes32 := mload(add(messageBytes, 116)) // 84 + 32 = 116
+    }
+    address destinationCaller = address(uint160(uint256(destinationCallerBytes32)));
+
+    return destinationCaller == legacyPool;
+  }
+
+  /// @notice Converts a legacy sourcePoolData struct into a new format that can be used to release or mint USDC on the
+  /// previous pool. This is necessary because the sourcePoolData is stored in a different format in the previous pool,
+  /// and must be in a properly decodable format.
+  /// @param releaseOrMintIn The releaseOrMintIn struct to generate a new struct for.
+  /// @return newReleaseOrMintIn The new releaseOrMintIn struct.
+  function _generateNewReleaseOrMintIn(
+    Pool.ReleaseOrMintInV1 calldata releaseOrMintIn
```

Comment:
Why does it take and return ReleaseOrMintInV1 when only sourcePoolData is relevant?